From 220af157ace03d5c00abb6a760931eb1abae72e5 Mon Sep 17 00:00:00 2001
From: Don Hiatt <don.hiatt@intel.com>
Date: Tue, 5 May 2020 15:39:07 -0700
Subject: [PATCH] drm/i915: Add rc6 enable/disable

Add the ability to enable/disable rc6.

To enable rc6:
   echo 1 > /sys/class/drm/card0/gt_rc6_enable

To disable rc6:
   echo 0 > /sys/class/drm/card0/gt_rc6_enable

To display the rc6 state:
   cat /sys/class/drm/card0/gt_rc6_enable

Signed-off-by: Don Hiatt <don.hiatt@intel.com>
Reviewed-by: Andi Shyti <andi.shyti@intel.com>
Signed-off-by: James Xiong <james.xiong@intel.com>
---
 drivers/gpu/drm/i915/gt/intel_rc6.c | 14 +++----
 drivers/gpu/drm/i915/gt/intel_rc6.h |  3 ++
 drivers/gpu/drm/i915/i915_sysfs.c   | 63 ++++++++++++++++++++---------
 3 files changed, 54 insertions(+), 26 deletions(-)

diff --git a/drivers/gpu/drm/i915/gt/intel_rc6.c b/drivers/gpu/drm/i915/gt/intel_rc6.c
index c3155ee586892..fd3fc2bb55fc4 100644
--- a/drivers/gpu/drm/i915/gt/intel_rc6.c
+++ b/drivers/gpu/drm/i915/gt/intel_rc6.c
@@ -500,14 +500,14 @@ static bool rc6_supported(struct intel_rc6 *rc6)
 	return true;
 }
 
-static void rpm_get(struct intel_rc6 *rc6)
+void intel_rc6_rpm_get(struct intel_rc6 *rc6)
 {
 	GEM_BUG_ON(rc6->wakeref);
 	pm_runtime_get_sync(rc6_to_i915(rc6)->drm.dev);
 	rc6->wakeref = true;
 }
 
-static void rpm_put(struct intel_rc6 *rc6)
+void intel_rc6_rpm_put(struct intel_rc6 *rc6)
 {
 	GEM_BUG_ON(!rc6->wakeref);
 	pm_runtime_put(rc6_to_i915(rc6)->drm.dev);
@@ -552,7 +552,7 @@ void intel_rc6_init(struct intel_rc6 *rc6)
 	int err;
 
 	/* Disable runtime-pm until we can save the GPU state with rc6 pctx */
-	rpm_get(rc6);
+	intel_rc6_rpm_get(rc6);
 
 	if (!rc6_supported(rc6))
 		return;
@@ -575,7 +575,7 @@ void intel_rc6_sanitize(struct intel_rc6 *rc6)
 	memset(rc6->prev_hw_residency, 0, sizeof(rc6->prev_hw_residency));
 
 	if (rc6->enabled) { /* unbalanced suspend/resume */
-		rpm_get(rc6);
+		intel_rc6_rpm_get(rc6);
 		rc6->enabled = false;
 	}
 
@@ -618,7 +618,7 @@ void intel_rc6_enable(struct intel_rc6 *rc6)
 		return;
 
 	/* rc6 is ready, runtime-pm is go! */
-	rpm_put(rc6);
+	intel_rc6_rpm_put(rc6);
 	rc6->enabled = true;
 }
 
@@ -666,7 +666,7 @@ void intel_rc6_disable(struct intel_rc6 *rc6)
 	if (!rc6->enabled)
 		return;
 
-	rpm_get(rc6);
+	intel_rc6_rpm_get(rc6);
 	rc6->enabled = false;
 
 	__intel_rc6_disable(rc6);
@@ -683,7 +683,7 @@ void intel_rc6_fini(struct intel_rc6 *rc6)
 		i915_gem_object_put(pctx);
 
 	if (rc6->wakeref)
-		rpm_put(rc6);
+		intel_rc6_rpm_put(rc6);
 }
 
 static u64 vlv_residency_raw(struct intel_uncore *uncore, const i915_reg_t reg)
diff --git a/drivers/gpu/drm/i915/gt/intel_rc6.h b/drivers/gpu/drm/i915/gt/intel_rc6.h
index e119ec4a0bcc9..f638db5e2b240 100644
--- a/drivers/gpu/drm/i915/gt/intel_rc6.h
+++ b/drivers/gpu/drm/i915/gt/intel_rc6.h
@@ -21,6 +21,9 @@ void intel_rc6_sanitize(struct intel_rc6 *rc6);
 void intel_rc6_enable(struct intel_rc6 *rc6);
 void intel_rc6_disable(struct intel_rc6 *rc6);
 
+void intel_rc6_rpm_get(struct intel_rc6 *rc6);
+void intel_rc6_rpm_put(struct intel_rc6 *rc6);
+
 u64 intel_rc6_residency_ns(struct intel_rc6 *rc6, i915_reg_t reg);
 u64 intel_rc6_residency_us(struct intel_rc6 *rc6, i915_reg_t reg);
 
diff --git a/drivers/gpu/drm/i915/i915_sysfs.c b/drivers/gpu/drm/i915/i915_sysfs.c
index 834296edb1286..06b46931bc920 100644
--- a/drivers/gpu/drm/i915/i915_sysfs.c
+++ b/drivers/gpu/drm/i915/i915_sysfs.c
@@ -59,23 +59,6 @@ static u32 calc_residency(struct drm_i915_private *dev_priv,
 	return DIV_ROUND_CLOSEST_ULL(res, 1000);
 }
 
-static ssize_t rc6_enable_show(struct device *kdev,
-			       struct device_attribute *attr, char *buf)
-{
-	struct drm_i915_private *dev_priv = kdev_minor_to_i915(kdev);
-	unsigned int mask;
-
-	mask = 0;
-	if (HAS_RC6(dev_priv))
-		mask |= BIT(0);
-	if (HAS_RC6p(dev_priv))
-		mask |= BIT(1);
-	if (HAS_RC6pp(dev_priv))
-		mask |= BIT(2);
-
-	return sysfs_emit(buf, "%x\n", mask);
-}
-
 static ssize_t rc6_residency_ms_show(struct device *kdev,
 				     struct device_attribute *attr, char *buf)
 {
@@ -108,14 +91,12 @@ static ssize_t media_rc6_residency_ms_show(struct device *kdev,
 	return sysfs_emit(buf, "%u\n", rc6_residency);
 }
 
-static DEVICE_ATTR_RO(rc6_enable);
 static DEVICE_ATTR_RO(rc6_residency_ms);
 static DEVICE_ATTR_RO(rc6p_residency_ms);
 static DEVICE_ATTR_RO(rc6pp_residency_ms);
 static DEVICE_ATTR_RO(media_rc6_residency_ms);
 
 static struct attribute *rc6_attrs[] = {
-	&dev_attr_rc6_enable.attr,
 	&dev_attr_rc6_residency_ms.attr,
 	NULL
 };
@@ -366,9 +347,51 @@ static ssize_t gt_min_freq_mhz_store(struct device *kdev,
 	return ret ?: count;
 }
 
+static ssize_t gt_rc6_enable_show(struct device *kdev, struct device_attribute *attr, char *buf)
+{
+	struct drm_i915_private *dev_priv = kdev_minor_to_i915(kdev);
+	struct intel_gt *gt = to_gt(dev_priv);
+
+	return scnprintf(buf, PAGE_SIZE, "%d\n", gt->rc6.enabled);
+}
+
+static ssize_t gt_rc6_enable_store(struct device *kdev,
+				   struct device_attribute *attr,
+				   const char *buf, size_t count)
+{
+	struct drm_i915_private *dev_priv = kdev_minor_to_i915(kdev);
+	struct intel_gt *gt = to_gt(dev_priv);
+
+	intel_wakeref_t wakeref;
+	ssize_t ret;
+	u32 val;
+
+	ret = kstrtou32(buf, 0, &val);
+	if (ret)
+		return ret;
+	wakeref = intel_runtime_pm_get(gt->uncore->rpm);
+	if (val) {
+		if (gt->rc6.enabled)
+			goto unlock;
+		if (!gt->rc6.wakeref)
+			intel_rc6_rpm_get(&gt->rc6);
+		intel_rc6_enable(&gt->rc6);
+		intel_rc6_unpark(&gt->rc6);
+	} else {
+		intel_rc6_disable(&gt->rc6);
+
+		if (gt->rc6.wakeref)
+			intel_rc6_rpm_put(&gt->rc6);
+	}
+unlock:
+	intel_runtime_pm_put(gt->uncore->rpm, wakeref);
+	return count;
+}
+
 static DEVICE_ATTR_RO(gt_act_freq_mhz);
 static DEVICE_ATTR_RO(gt_cur_freq_mhz);
 static DEVICE_ATTR_RW(gt_boost_freq_mhz);
+static DEVICE_ATTR_RW(gt_rc6_enable);
 static DEVICE_ATTR_RW(gt_max_freq_mhz);
 static DEVICE_ATTR_RW(gt_min_freq_mhz);
 
@@ -402,6 +425,7 @@ static const struct attribute * const gen6_attrs[] = {
 	&dev_attr_gt_act_freq_mhz.attr,
 	&dev_attr_gt_cur_freq_mhz.attr,
 	&dev_attr_gt_boost_freq_mhz.attr,
+	&dev_attr_gt_rc6_enable.attr,
 	&dev_attr_gt_max_freq_mhz.attr,
 	&dev_attr_gt_min_freq_mhz.attr,
 	&dev_attr_gt_RP0_freq_mhz.attr,
@@ -414,6 +438,7 @@ static const struct attribute * const vlv_attrs[] = {
 	&dev_attr_gt_act_freq_mhz.attr,
 	&dev_attr_gt_cur_freq_mhz.attr,
 	&dev_attr_gt_boost_freq_mhz.attr,
+	&dev_attr_gt_rc6_enable.attr,
 	&dev_attr_gt_max_freq_mhz.attr,
 	&dev_attr_gt_min_freq_mhz.attr,
 	&dev_attr_gt_RP0_freq_mhz.attr,
-- 
2.25.1

